src/compose/ComposeBox.js

```diff
@@ -155,32 +224,31 @@ export default function ComposeBox(props: Props): Node {
     either: false,
   });
 
-  // The topic input is currently uncontrolled, for performance concerns;
-  // see #2738. That means if you change this state, it won't cause the
-  // input value to change; callers should call updateTextInput for that.
-  // But if callers do that, they should *also* update this state, because
-  // it's our best read of what the actual value is, so it needs to be
-  // up-to-date. That's what setTopicInputValue (far below) is for; it does
-  // both.
-  // TODO(?): Make custom Hook to encapsulate this uncontrolled-input logic.
-  const [topicInputState, setTopicInputState] = useState<{| value: string |}>({
-    value: initialTopic ?? (isTopicNarrow(narrow) ? topicOfNarrow(narrow) : ''),
+  const [
+    topicInputState,
+    setTopicInputValue,
+    setTopicInputSelection /* eslint-disable-line no-unused-vars */,
+    {
+      _handleValueChange: _handleTopicValueChange,
+      _handleSelectionChange: _handleTopicSelectionChange,
+    },
+  ] = useUncontrolledInput({
+    ref: topicInputRef,
+    initialValue: initialTopic ?? (isTopicNarrow(narrow) ? topicOfNarrow(narrow) : ''),
   });
 
-  // The message input is currently uncontrolled, for performance concerns;
-  // see #2738. That means if you change this state, it won't cause the
-  // input value to change; callers should call updateTextInput for that.
-  // But if callers do that, they should *also* update this state, because
-  // it's our best read of what the actual value is, so it needs to be
-  // up-to-date. That's what setMessageInputValue (far below) is for; it does
-  // both.
-  // TODO(?): Make custom Hook to encapsulate this uncontrolled-input logic.
-  const [messageInputState, setMessageInputState] = useState<{|
-    value: string,
-    selection: InputSelection,
-  |}>({
-    value: initialMessage ?? '',
-    selection: { start: 0, end: 0 },
+  const [
+    messageInputState,
+    setMessageInputValue,
+    setMessageInputSelection /* eslint-disable-line no-unused-vars */,
+    {
+      _handleValueChange: _handleMessageValueChange,
+      _handleSelectionChange: _handleMessageSelectionChange,
+    },
+  ] = useUncontrolledInput({
```

Comment:
I think we can reduce the number of separate names for things here by doing something like:
* This object returned by `useControlledInput` calls its properties just `onChangeText` and  `onSelectionChange` -- the same names as the `TextInput` props they're supposed to be passed to
* Instead of destructuring the object here, we give it a name like `messageInputCallbacks`
* When constructing the `TextInput` element, we say `onChangeText={messageInputCallbacks.onChangeText}` and so on
  * Or perhaps even just `{...messageInputCallbacks}`